Jetpack Boost: Refactor Image Size Analysis

pyronaur opened the pull request:
## Proposed changes:
Refactor the entire image guide UI to React.js

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
N/A

## Does this pull request change what data or activity we track or use?
N/A

## Testing instructions:
Check out the branch, build all the things, play use the image size analysis module and compare to the current trunk _(don't compare to svelte version because transitions don't exist in our React app at the moment - we don't have a way to use transitions right now)_

pyronaur committed:
MultiProgress.svelte

pyronaur committed:
rm MultiProgress.svelte

pyronaur committed:
Add icon className

pyronaur committed:
Fix spinner class

pyronaur committed:
cleanup import

pyronaur committed:
rm RecommendationsMeta

pyronaur committed:
rm ConditionalLink.svelte

pyronaur committed:
Convert svg files to react components

pyronaur committed:
Migrate the whole ISA UI to React

pyronaur committed:
Merge branch 'trunk' into boost-react/recommendations-page

pyronaur committed:
Labels are not stores

pyronaur committed:
💔

pyronaur committed:
Merge branch 'trunk' into boost-react/recommendations-page

pyronaur committed:
Fix Table component

pyronaur committed:
Fix image-missing-row

pyronaur committed:
Unused createInterpolateElement

pyronaur committed:
Fix table & rows

pyronaur committed:
Fix button component UI

pyronaur committed:
Remove mock api

pyronaur committed:
Fix hover state

pyronaur committed:
remove ignored code

pyronaur committed:
Shorten mount

pyronaur committed:
Remove manual refreshing

pyronaur committed:
Clean-up

pyronaur committed:
Simplify Table component

pyronaur committed:
Strict mode

pyronaur committed:
strict-ts-fix

pyronaur committed:
TS: isaGroupLabel

pyronaur committed:
TSFix

pyronaur committed:
Merge branch 'trunk' into boost-react/recommendations-page

pyronaur committed:
jsx is unnecessary

pyronaur committed:
Strict ImageData with no fallback

pyronaur committed:
Fix row data

pyronaur committed:
Remove derived stores

pyronaur committed:
Fix expandable type errors

pyronaur committed:
Fix translations comment not being picked up

pyronaur committed:
Prepare for isa-summary refactor

pyronaur committed:
Fix types

pyronaur committed:
Simplify error message

pyronaur committed:
De-dupe issue count

pyronaur committed:
Fix image counting bugs

pyronaur committed:
Split out types from logic

pyronaur committed:
Move all Zod types to "types.ts"

pyronaur committed:
Remove "resetIsaQuery"

It's unclear at the moment if this is going to be necessary, removing this for now

pyronaur committed:
Refactor ISA Types

pyronaur committed:
Fix pagination navigation

pyronaur committed:
Only export types as public interafce from types file

pyronaur committed:
Add comments

pyronaur committed:
Improve isa group label naming

pyronaur committed:
Counts are fixed now

pyronaur committed:
isGroupLabels -> isaGroupKeys

pyronaur committed:
Add `useIsaData`

pyronaur committed:
Add `useIsaReport`

pyronaur committed:
Accept string as group name

pyronaur committed:
Convert ImageSizeAnalysis root component to react

pyronaur committed:
TODO: Figure out navigation with DataSync

pyronaur committed:
Add ACTION to DataSync

pyronaur committed:
WIP: useDataSyncAction

pyronaur committed:
Restore Pagination in React

pyronaur committed:
Add missing comment

pyronaur committed:
Remove logs

pyronaur committed:
Restore image fixer: WIP

pyronaur committed:
Remove {details}

pyronaur committed:
TEMPORARY HACK: Manually register DS Action Endpoint

pyronaur committed:
All params are required

pyronaur committed:
Removed experiment code

pyronaur committed:
Working useDataSyncAction() prototype

pyronaur committed:
DataSync Example: useImageFixer()

pyronaur committed:
changelogger

thingalon commented on projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:
```diff
@@ -125,3 +126,118 @@ export function useDataSync<
+ *
+ *
+ */
+export function useDataSyncAction< ActionData extends RequestParams >() {
+	/*
+	 * Yes, this is a function that returns a function.
+	 * Also yes - this is dumb.
+	 * But TypeScript still doesn't allow partial type inference.
+	 *
+	 * This is what happens when you rush development
+	 * - you get legacy code that can't be fixed for years.
+	 * @see https://github.com/microsoft/TypeScript/issues/26242
+	 */
+
+	return function <
+		Schema extends z.ZodSchema,
+		State extends z.infer< Schema >,
+		Key extends string,
+		ActionName extends string,
+		ActionSchema extends z.ZodSchema,
+		ActionResult extends z.infer< ActionSchema >,
+		Config extends DataSyncConfig< ActionSchema, ActionData >[ 'mutation' ],
+	>(
+		namespace: string,
+		key: Key,
+		name: ActionName,
+		stateSchema: Schema,
+		actionSchema: ActionSchema,
+		callback: ( result: ActionResult, currentValue: State ) => void | State,
+		config: Config = {} as Config,
+		params: Record< string, string | number > = {}
+	) {
```
"No," I said to myself. "Surely this isn't the only way to deal with complex typing issues like this," I said.
I fell down the rabbit hole of looking into this and 😱 it _is_ how you do it. I shall never be the same again.

Aside from this ugly corner, which I acknowledge is a workaround for TypeScript, this is looking like a great solution 👍

pyronaur committed:
Add jetpack_boost_register_action() to DataSync PHP side

pyronaur committed:
monofix

pyronaur committed:
Refactor useDataSyncAction to accept a configuration object

pyronaur committed:
Get rid of double function call

pyronaur committed:
Add window error handling

pyronaur committed:
Clearer action names

pyronaur committed:
Fix isa fix

pyronaur committed:
Optimistic update handler

pyronaur committed:
Parametirize useImageFixer()

pyronaur committed:
Make callbacks optional and rename config to mutationOptions

pyronaur committed:
Cleanup

pyronaur committed:
Clean up custom options

pyronaur committed:
Document useDataSyncAction() config options

pyronaur committed:
Cleanup

pyronaur committed:
Add changelogger entries

pyronaur committed:
Remove isaReportDS

pyronaur committed:
Restore polling

pyronaur committed:
Convert ISA start to DataSync actions

pyronaur committed:
Load initialData later

pyronaur committed:
Default params

pyronaur committed:
Add todos and an export

pyronaur committed:
Simplify pagination

pyronaur committed:
Cleanup

pyronaur committed:
Merge branch 'trunk' into boost-react/recommendations-page

# Conflicts:
#	projects/plugins/boost/app/assets/src/js/Main.svelte
#	projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx
#	projects/plugins/boost/app/assets/src/js/pages/index/Index.svelte
#	projects/plugins/boost/app/assets/src/js/pages/purchase-success/purchase-success.tsx
#	projects/plugins/boost/app/assets/src/js/svg/right-arrow.tsx
#	projects/plugins/boost/wp-js-data-sync.php

pyronaur committed:
Merge requestNewReport

pyronaur committed:
Deal with post-merge fun

pyronaur committed:
Restore ISA in main

pyronaur committed:
Pagination & Navigation

pyronaur committed:
Fix CSS selectors

pyronaur committed:
Monofix

pyronaur committed:
Add changelog

pyronaur committed:
Remove unused code

pyronaur committed:
Harden get()

pyronaur committed:
Add a todo

haqadn commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx:
```diff
@@ -1,7 +1,238 @@
+											'Congratulations; no issues found after scanning your %d most recent pages.',
+											'jetpack-boost'
+										),
+										scannedPages
+									) }
+								</div>
+							) }
+							<button
+								type="button"
+								className="components-button is-link"
+								onClick={ handleAnalyzeClick }
+								disabled={ requestingReport }
+							>
+								<RefreshIcon />
+								{ __( 'Analyze again', 'jetpack-boost' ) }
+							</button>
```
Can we use the RNA button for this?

pyronaur replied:
👍  - maybe that can be a follow-up? This is already 2.4k loc.

dilirity replied:
Will be done in follow-up PR.

haqadn commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/button/button.tsx:
```diff
@@ -0,0 +1,49 @@
+	} );
+
+	return href ? (
+		<a href={ href } onClick={ onClick } className={ buttonClassNames } style={ buttonStyle }>
+			{ children }
+		</a>
```
I like how this has a `<a>` based on href. We could make this button component not limited to ISA. Or even better,  the RNA button component could be improved to use the `<a>` depending on the `href` prop.

pyronaur replied:
👍

dilirity replied:
Will be done in follow-up PR.

haqadn commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/button/button.tsx:
```diff
@@ -0,0 +1,49 @@
+			{ children }
+		</a>
+	) : (
+		<button
+			onClick={ onClick }
+			className={ buttonClassNames }
+			disabled={ disabled }
+			style={ buttonStyle }
+		>
+			{ children }
+		</button>
```
This can use the RNA component.

pyronaur replied:
👍  - maybe that can be a follow-up? This is already 2.4k loc.

dilirity replied:
Will be done in follow-up PR.

haqadn commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/conditional-link/conditional-link.tsx:
```diff
@@ -0,0 +1,35 @@
+	if ( isLink ) {
+		return (
+			// eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions
+			<a onClick={ handleClick } { ...rest }>
+				{ children }
+			</a>
```
Based on the usage of the ConditionalLink, seems like the route is passed via `to` prop. I am not sure how that works with `<a>`. But, the `Link` component from `react-router-dom` does use the `to` prop. It also puts proper `href` for the rendered link.

pyronaur replied:
👍

dilirity replied:
Will be done in follow-up PR.

haqadn commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx:
```diff
@@ -1,7 +1,238 @@
+const _RecommendationsMeta: React.FC< RecommendationsMetaProps > = ( { isCdnActive } ) => {
+	const [ requestingReport, setRequestingReport ] = useState< boolean >( false );
+	const [ errorCode, setErrorCode ] = useState< number | undefined >( undefined );
+	const [ status, setStatus ] = useState< ISAStatus | undefined >( undefined );
+	const [ groups, setGroups ] = useState< Record< string, IsaCounts > >( {} );
+	const [ scannedPages, setScannedPages ] = useState< number >( 0 );
+	const [ totalIssues, setTotalIssues ] = useState< number >( 0 );
+	const [ errorMessage, setErrorMessage ] = useState< string | undefined >( undefined );
+	const [ errorSuggestion, setErrorSuggestion ] = useState< string | undefined >( undefined );
+	const [ waitNotice, setWaitNotice ] = useState< string | undefined >( undefined );
+	const [ showCDNRecommendation, setShowCDNRecommendation ] = useState< boolean >( false );
+	const [ { data: isaReport } ] = useIsaReport();
+	const requestNewReport = useImageAnalysisRequest();
+
+	useEffect( () => {
+		setStatus( isaReport?.status );
+		setGroups( isaReport?.groups || {} );
+		setScannedPages( scannedPagesCount( isaReport?.groups || {} ) );
```
It seems like these states are used like variables. Can we skip the states and use variables?

```suggestion
	const [ totalIssues, setTotalIssues ] = useState< number >( 0 );
	const [ errorMessage, setErrorMessage ] = useState< string | undefined >( undefined );
	const [ errorSuggestion, setErrorSuggestion ] = useState< string | undefined >( undefined );
	const [ waitNotice, setWaitNotice ] = useState< string | undefined >( undefined );
	const [ showCDNRecommendation, setShowCDNRecommendation ] = useState< boolean >( false );
	const [ { data: isaReport } ] = useIsaReport();
	const requestNewReport = useImageAnalysisRequest();

	const status = isaReport?.status;
	const groups = isaReport?.groups || {};
	const scannedPages = scannedPagesCount( isaReport?.groups || {} )

	useEffect( () => {
```

pyronaur replied:
Possibly, will circle back on this. 👍

haqadn commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx:
```diff
@@ -1,7 +1,238 @@
+
+const RecommendationsMeta = ( { isCdnActive }: { isCdnActive: boolean } ) => {
+	return (
+		<DataSyncProvider>
+			<_RecommendationsMeta isCdnActive={ isCdnActive } />
+		</DataSyncProvider>
```
There is a top level `DataSyncProvider` already added to `main.tsx`

pyronaur replied:
👍

haqadn commented on projects/plugins/boost/app/modules/image-guide/src/initialize.ts:
```diff
@@ -133,42 +133,45 @@ function findContainer( image: MeasurableImage ): HTMLElement | undefined {
-				},
-			};
-		}
+			// Don't create new entry for Svelte component configuration.
+			// Use the index in the array as a unique identifier.
```
The comment seems outdated

haqadn commented on projects/plugins/boost/app/modules/image-guide/src/initialize.ts:
```diff
@@ -133,42 +133,45 @@ function findContainer( image: MeasurableImage ): HTMLElement | undefined {
+			const store = new MeasurableImageStore( image );
+			stores.push( store );
+
+			// If there's only one image, assume a new Svelte component needs to be created.
```
Putting aside that it says svelte 😛 , I don't understand creating the component part. Maybe rephrase the comment?

pyronaur replied:
This is not a part of jetpack boost UI - I think the changes here are only caused by linter.

haqadn commented on projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Action_Fix.php:
```diff
@@ -0,0 +1,66 @@
+	 */
+	public function handle( $data, $request ) {
+
+		// @TODO: Add a proper feature flag for this instead of just checking if priority support available.
```
I think the @TODO comment is outdated.

pyronaur replied:
No idea, I just moved it

dilirity replied:
I'm also confused about this (though it's before this PR as Nauris mentioned). Is this for feature gating the ISA fixer?

haqadn commented on projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Entry.php:
```diff
@@ -4,26 +4,25 @@
-	private $page         = 1;
-	private $group        = 'all';
-	private $search_query = '';
+		// @REACT-TODO: This could be moved to DataSync actions
```
What can be moved to DS actions?

pyronaur replied:
Pagination code. It's an experiment I'll think about in 2024.

haqadn reviewed (commented):
I am afraid to say I reviewed it because it is huge 🤯. I have noted some of the things I noticed.

The build failed so I couldn't test it yet.

pyronaur commented:
Try installing and rebuilding all the things - the build is building for me.

haqadn commented:
> Try installing and rebuilding all the things - the build is building for me.

Build fails on both the github workflow and my machine.

```
[validate-ts] app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx:46:5 - error TS2345: Argument of type 'unknown' is not assignable to parameter of type 'SetStateAction<number>'.
[validate-ts] 
[validate-ts] 46     Object.values( isaReport.groups ).reduce( ( total, group ) => total + group.issue_count, 0 )
[validate-ts]        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[validate-ts] 
[validate-ts] app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx:46:81 - error TS2339: Property 'issue_count' does not exist on type 'unknown'.
[validate-ts] 
[validate-ts] 46     Object.values( isaReport.groups ).reduce( ( total, group ) => total + group.issue_count, 0 )
[validate-ts]                                                                                    ~~~~~~~~~~~
[validate-ts] 
[validate-ts] 
[validate-ts] Found 2 errors in the same file, starting at: app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx:46
[validate-ts] 
[validate-ts]  ELIFECYCLE  Command failed with exit code 2.
```

dilirity commented:
Can confirm, build/run dev fails just like @haqadn's

dilirity committed:
Fix build failing

dilirity commented:
Build is now fixed.

EDIT:
No need to build all. Just:
- `jetpack build js-packages/react-data-sync-client`
- `jetpack build packages/wp-js-data-sync`
- `jetpack build plugins/boost`

dilirity commented on projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:
```diff
@@ -63,7 +70,7 @@ export function useDataSync<
 ): DataSyncHook< Schema, Value > {
 	const datasync = new DataSync( namespace, key, schema );
-	const queryKey = [ key, ...Object.values( params ) ];
+	const queryKey = [ key, ...Object.values( params ).sort() ];
```
Why is `sort()` needed here?

pyronaur replied:
You got it

dilirity commented on projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:
```diff
@@ -125,3 +132,162 @@ export function useDataSync<
+	StateSchema extends z.ZodSchema,
+	ActionSchema extends z.ZodSchema,
+	ActionResult extends z.infer< ActionSchema >,
+	CurrentState extends z.infer< StateSchema >,
```
`extend` all the things!

dilirity commented on projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:
```diff
@@ -125,3 +132,162 @@ export function useDataSync<
+	CurrentState extends z.infer< StateSchema >,
+> = {
+	/**
+	 * The project namespace, fore example: 'jetpack_boost_ds'
```
`fore` - thank you for the examples in the comments <3

dilirity commented on projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts:
```diff
@@ -125,3 +132,162 @@ export function useDataSync<
+	CurrentState
+> ) {
+	// @REACT-TODO: order sensitive bug is hiding in Object.values
+	// This `sort` of fixes it, but I'd like a more elegant solution.
```
Oooh 🔩s! That's why you need the `sort()`.

dilirity commented on projects/plugins/boost/app/modules/image-guide/src/initialize.ts:
```diff
@@ -133,42 +133,45 @@ function findContainer( image: MeasurableImage ): HTMLElement | undefined {
-			console.error( `Image has no parent`, image.node );
-			return acc;
-		}
+	const componentConfiguration = measuredImages.reduce(
+		( acc, image ) => {
```
Should formatting like this happen or is this how it's suppose to look like? It appears it introduced unnecessary changes in the PR.

pyronaur replied:
Auto-formatter did this, looks like I committed this accidentally, but I guess we can leave it - it's just a formatting change.

dilirity commented on projects/plugins/boost/app/assets/src/js/svg/warning-outline.tsx:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
Thish ishn't ushed.

dilirity commented on projects/plugins/boost/app/assets/src/js/svg/refresh.tsx:
```diff
@@ -1,15 +1,17 @@
-	</svg>
-);
-
+import React from 'react';
```
Wait, why is this necessary?

dilirity commented on projects/plugins/boost/app/assets/src/js/svg/chevron-left.tsx:
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
```
Here it is again 👻

dilirity commented on projects/plugins/boost/app/assets/src/js/svg/checkbox.tsx:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
```
And again 😲

dilirity commented on projects/plugins/boost/app/assets/src/js/pages/image-size-analysis/image-size-analysis.tsx:
```diff
@@ -0,0 +1,76 @@
+	return (
+		<DataSyncProvider>
+			<_ImageSizeAnalysis { ...props } />
+		</DataSyncProvider>
```
Shouldn't this be removed, since we already have a provider wrapped around `Main`?

dilirity commented on projects/plugins/boost/app/assets/src/js/features/ui/spinner/spinner.tsx:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import styles from './spinner.module.scss';
+
+interface SpinnerProps {
+	size?: string;
+	lineWidth?: string;
+}
```
Sometimes we use interface, sometimes type, sometimes we use

```js
const Spinner: React.FC< SpinnerProps > = ( { size = '1.4rem', lineWidth = '2px' } ) => {
```

or

```js
const SettingsPage = ( { children }: SettingsPageProps ) => {
```

I couldn't figure out what the difference is, other than the React.FC is something old.

Anyway - this is more just me wondering about this. Let's not change things because of that 😅

dilirity commented on projects/plugins/boost/app/assets/src/js/features/ui/spinner/spinner.module.scss:
```diff
@@ -0,0 +1,14 @@
+	to {
+	  transform: rotate(360deg);
+	}
+  }
```
I know it's nit-picky, but there's mixed indentation.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/ui/thumbnail/thumbnail.tsx:
```diff
@@ -0,0 +1,28 @@
+const Thumbnail: React.FC< ThumbnailProps > = ( { url, title, width, height } ) => {
+	const thumbnailStyle = {
+		'--thumbnail-size': `${ width }px`,
+	} as React.CSSProperties;
```
TIL about `React.CSSProperties`.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/ui/thumbnail/thumbnail.tsx:
Styles from `thumbnail.scss` can be moved next to this file.

It also appears that `.jb-thumbnail__image--missing` is not used anywhere?

dilirity replied:
Will be done in a follow up PR.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/ui/row-title/row-title.tsx:
`row-title.scss` can be moved next to this file.

dilirity replied:
Will be done in a follow up PR.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/ui/pill/pill.tsx:
`pill.scss` can be moved next to this file.

dilirity replied:
Will be done in a follow up PR.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/ui/pagination-arrow/pagination-arrow.tsx:
```diff
@@ -0,0 +1,38 @@
+		return <span className="jb-pagination__page jb-pagination__page--inactive">{ children }</span>;
+	}
+	return (
+		<a href={ `#/image-size-analysis/${ group }/${ page }` } className="jb-pagination__page">
```
Shouldn't we be using some kind of react magic instead of plain anchors to navigate? Just wondering.

pyronaur replied:
I think the react router is the magic - it listens for window location href changes.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/ui/pagination-arrow/pagination-arrow.tsx:
I'm a bit confused by the name of the file and the name of the component. Shouldn't it be link? Arrow is just cosmetics, no?

dilirity replied:
It was moved to the main Pagination component file as it's used only there.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/ui/pagination-arrow/pagination-arrow.module.scss:
I guess you started moving the styles to their own file but didn't finish.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/tabs/tabs.tsx:
`tabs.scss` can be moved next to this file.

pyronaur replied:
added a todo to the board

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/tabs/tabs.tsx:
```diff
@@ -0,0 +1,107 @@
+					{ /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */ }
+					<div className="jb-dropdown__head-bar" onClick={ onClickDropdown }>
+						{ getGroupLabel( activeGroupKey ) }
+						{ currentTab && currentTab?.issue_count > 0 && (
```
I don't think the question mark necessary if there's a check for `currentTab` first.

pyronaur replied:
You're right

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/tabs/tabs.tsx:
```diff
@@ -0,0 +1,107 @@
+	};
+
+	const onClickDropdown = () => {
+		setDropdownOpen( ! dropdownOpen );
```
I can't remember why in the Joy of React, the guy mentioned that he likes to use the callback and invert the value in the callback, instead of inverting the state variable.

Like

```js
setDropdownOpen( value => ! value );
```

Do you remember why he said he likes doing that over this way?

pyronaur replied:
From the joy of react course: 

> Most of the time, we don't need to use the callback syntax. This alternative syntax is only necessary when it's possible for state variables to grow stale.

So I think this is fine in this case - it's jut local state that keeps track of dropdown state

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/tabs/tabs.module.scss:
Oh, I guess tabs styles can't be moved here then.

pyronaur replied:
You probably meant can be moved here. Yes they should, but probably not in this 2k PR 😉

dilirity commented:
If you request an analysis and quickly click the view report button, you're presented with this:

![CleanShot 2023-12-22 at 14 12 45](https://github.com/Automattic/jetpack/assets/11799079/607d67f9-00d0-4b18-a353-249a01233c22)

dilirity commented on projects/packages/wp-js-data-sync/src/endpoints/class-action-endpoint.php:
```diff
@@ -0,0 +1,90 @@
+	 * @param $namespace
+	 * @param $route
+	 * @param $action_class
+	 */
```
Outdated comment.

What is `key`?

dilirity commented on projects/packages/wp-js-data-sync/src/endpoints/class-action-endpoint.php:
```diff
@@ -0,0 +1,90 @@
+		$this->action_class   = $action_class;
+		$this->rest_namespace = $namespace;
+		$this->route          = $key . '/action/' . $action_name;
+		$this->nonce          = new Authenticated_Nonce( "{$namespace}_{$this->route}_action" );
```
I just noticed, that this class is also present in the `Async_Option` package.

pyronaur replied:
Oh, that's just the inspect plugin that's the origin of DataSync. It needs to get updated, but that's not a problem for this PR.

dilirity commented on projects/packages/wp-js-data-sync/src/endpoints/class-action-endpoint.php:
```diff
@@ -0,0 +1,90 @@
+				)
+			);
+		} catch ( \Error $e ) {
+			return rest_ensure_response(
+				new \WP_Error( 500, $e->getMessage(), array( 'status' => 500 ) )
+			);
```
I always thought you can just return a new WP_Error instance.

pyronaur replied:
You might be able to, I haven't looked in too deep here, I just now that `rest_ensure_response` ensures that it's ok, so I prefer to stay safe.

dilirity commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/hero/hero.tsx:
```diff
@@ -24,27 +25,26 @@ export const Hero = ( {
 								_n(
 									'%d Image Recommendation',
 									'%d Image Recommendations',
-									totalIssueCount,
```
Thanks for fixing this 😅

dilirity reviewed (dismissed):
Some things I noticed.

pyronaur committed:
Improve the comment

pyronaur committed:
fore to for

pyronaur committed:
Remove unnecessary `?`

pyronaur committed:
Remove unused scss file

pyronaur committed:
No need to re-use pagination-arrow

pyronaur committed:
Use Link in Pagination

pyronaur committed:
Fix pagination using <Link> component

pyronaur committed:
Merge in Conditionallink

pyronaur committed:
Call it MaybeLink

pyronaur committed:
Update projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations-meta/recommendations-meta.tsx

Co-authored-by: Adnan Haque <3737780+haqadn@users.noreply.github.com>

pyronaur committed:
DataSyncProvider is unnecessary

pyronaur committed:
Remove unused imports

pyronaur committed:
Fix indentation

pyronaur committed:
Remove DataSyncProvider

pyronaur committed:
Remove unused React import

pyronaur committed:
Remove todo

pyronaur committed:
Prefix with underscores to make phpcs ignore the unused vars

pyronaur committed:
Log DataSync Errors

DataSync really needs better debugging tools for Zod, but this will have to do for now.

pyronaur committed:
Coerce numbers in IsaGlobal

pyronaur committed:
Use actions for pagination

pyronaur committed:
Reorganize so that the variable is available

pyronaur committed:
Catch and display the error

pyronaur committed:
Merge branch 'trunk' into boost-react/recommendations-page

# Conflicts:
#	projects/js-packages/react-data-sync-client/src/DataSync.ts
#	projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts
#	projects/packages/wp-js-data-sync/src/endpoints/class-action-endpoint.php
#	projects/plugins/boost/package.json

pyronaur committed:
Log errors

This should be improved further by logging only unhandled errors, we'll get there eventually.

pyronaur committed:
Clean-up the loading state

pyronaur committed:
Automatically refetch isaData when the report is loading

pyronaur committed:
DataSync Error class

pyronaur committed:
Refactor error handling

pyronaur committed:
TS Fix

pyronaur committed:
Add changelogger

Commit:
Fix compile error

thingalon commented on pnpm-lock.yaml:
```diff
@@ -2710,6 +2710,9 @@ importers:
         specifier: 18.2.33
         version: 18.2.33
+      '@types/react-router-dom':
+        specifier: 5.3.3
+        version: 5.3.3
```
That's an old version. Should we use 6?

pyronaur replied:
I just did `pnpm install ....` and that's what it installed. I think @haqadn is best geared to answer this because I think he was actually the one converting our router.

haqadn replied:
I have no idea why it installed this version. The `react-router-dom` version we are using is `6.21`. Maybe the types didn't change between these versions?

thingalon commented on projects/js-packages/react-data-sync-client/src/DataSyncError.ts:
```diff
@@ -0,0 +1,28 @@
+	) {
+		super( message );
+
+		/**
+		 * This makes `foo instanceof DataSyncError` work.
+		 * To make instanceof work correctly
+		 * set the prototype explicitly.
+		 *
+		 * @see https://stackoverflow.com/a/41102306/1015046
+		 */
```
Weeeeeeird 👻 

I've used instanceof to identify error types coming from critical css without any issue. Any idea why it needed a TypeScript workaround here?

pyronaur replied:
Hmm, this is just a file rename, I don't remember why, but `instanceof` didn't work before without this workaround, you even reviewed it back then and said TIL 😄  - I'd rather not change that in this mega-PR.

thingalon commented on projects/plugins/boost/app/assets/src/js/features/image-size-analysis/recommendations/row-types/table-row-hover/table-row-hover.tsx:
```diff
@@ -0,0 +1,56 @@
+import { recordBoostEventAndRedirect } from '$lib/utils/analytics';
+import { type IsaImage } from '$features/image-size-analysis';
+
+// @REACT-TODO: Move this to a DataSync Store.
```
Why?

pyronaur replied:
So that you can deliver a feature flag value from the back-end to the front-end.

thingalon replied:
Why? I'm not trying to be obnoxious here. I don't understand the value.

pyronaur replied:
So that we can feature-flag IsaFixer while it's not released. The UI has to change when ISA Fixer is active.

thingalon reviewed (approved):
Fixed a small compile bug. Seems to work. Let's merge and roll forward with it :)
